util/db/cluster.go

```diff
@@ -60,21 +60,31 @@ func (db *db) ListClusters(ctx context.Context) (*appv1.ClusterList, error) {
 	clusterList := appv1.ClusterList{
 		Items: make([]appv1.Cluster, 0),
 	}
+	inClusterServerAddressAllowed, err := db.settingsMgr.GetInClusterServerAddressAllowed()
+	if err != nil {
+		return nil, err
+	}
 	hasInClusterCredentials := false
 	for _, clusterSecret := range clusterSecrets {
 		cluster, err := secretToCluster(clusterSecret)
 		if err != nil {
 			log.Errorf("could not unmarshal cluster secret %s", clusterSecret.Name)
 			continue
 		}
-		clusterList.Items = append(clusterList.Items, *cluster)
 		if cluster.Server == appv1.KubernetesInternalAPIServerAddr {
-			hasInClusterCredentials = true
+			if inClusterServerAddressAllowed {
+				hasInClusterCredentials = true
+				clusterList.Items = append(clusterList.Items, *cluster)
+			} else {
+				log.Errorf("failed to add cluster %q to cluster list: in-cluster server address is disabled in Argo CD settings", cluster.Name)
+			}
+		} else {
+			if !hasInClusterCredentials {
+				clusterList.Items = append(clusterList.Items, *db.getLocalCluster())
+			}
+			clusterList.Items = append(clusterList.Items, *cluster)
```

Comment:
I think this is a bug. It seems it keeps adding the cluster over and over again.

Reply:
Updated the logic. Could you take another look when you get a chance?